### PR TITLE
TestHelpers: Move from coala-bears to coala

### PR DESCRIPTION
**Pull request by abhsag24**
bears should be testeable without coala-bears

Fixes https://github.com/coala-analyzer/coala-bears/issues/488

### CODE REVIEW

**Comment by sils1297** [comment]
yes, tests are needed

**Comment by Makman2** [comment]
@abhsag24 wanna give it a try again? :)

**Comment by abhsag24** [comment]
yeah i'll add the tests :D

File: coalib/testing/LocalBearTestHelper.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        Asserts that a check of the given lines with the given local bear
+        either yields or does not yield any results.
+
+        :param local_bear:       The local bear to check with.
+        :param lines:            The lines to check. (List of strings)
+        :param filename:         The filename, if it matters.
+        :param valid:            Whether the lines are valid or not.
+        :param force_linebreaks: Whether to append newlines at each line
+                                 if needed. (Bears expect a \\n for every line)
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Period at the end of the sentence inside the parantheses `... for every line.)`

File: coalib/testing/LocalBearTestHelper.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        Asserts that a check of the given lines with the given local bear does
+        yield exactly the given results.
+
+        :param local_bear:       The local bear to check with.
+        :param lines:            The lines to check. (List of strings)
+        :param results:          The expected list of results.
+        :param filename:         The filename, if it matters.
+        :param force_linebreaks: Whether to append newlines at each line
+                                 if needed. (Bears expect a \\n for every line)
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Same here

File: tests/testing/LocalBearTestHelperTest.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+from tests.test_bears.TestBear import (FirstTestBear,
+                                       SecondTestBear,
+                                       ThirdTestBear)
+from coalib.settings.Section import Section
+from coalib.settings.Setting import Setting
+from coalib.testing.LocalBearTestHelper import verify_local_bear, execute_bear
+
+
+files = (("Everything is invalid/valid/raises error",),)
```
Comment (hypothesist) [inline tests/testing/LocalBearTestHelperTest.py]: Why not initialize these inside the class `setUp` function? IMO that's what this is for. Or are you gonna create another class that will reuse this in the future?

File: coalib/testing/LocalBearTestHelper.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    :param settings:         A dictionary of keys and values (both string) from
+                             which settings will be created that will be made
+                             available for the tested bear.
+    :param force_linebreaks: Whether to append newlines at each line
+                             if needed. (Bears expect a \\n for every line)
+    :param create_tempfile:  Whether to save lines in tempfile if needed.
+    :param timeout:          The total time to run the test for.
+    :param tempfile_kwargs:  Kwargs passed to tempfile.mkstemp() if tempfile
+                             needs to be created.
+    :return:                 A unittest.TestCase object.
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Double backticks

File: coalib/testing/LocalBearTestHelper.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        either yields or does not yield any results.
+
+        :param local_bear:       The local bear to check with.
+        :param lines:            The lines to check. (List of strings)
+        :param filename:         The filename, if it matters.
+        :param valid:            Whether the lines are valid or not.
+        :param force_linebreaks: Whether to append newlines at each line
+                                 if needed. (Bears expect a \\n for every line)
+        :param create_tempfile:  Whether to save lines in tempfile if needed.
+        :param tempfile_kwargs:  Kwargs passed to tempfile.mkstemp().
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Backticks

File: coalib/testing/LocalBearTestHelper.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                             yield results.
+    :param filename:         The filename to use for valid and invalid files.
+    :param settings:         A dictionary of keys and values (both string) from
+                             which settings will be created that will be made
+                             available for the tested bear.
+    :param force_linebreaks: Whether to append newlines at each line
+                             if needed. (Bears expect a \\n for every line)
+    :param create_tempfile:  Whether to save lines in tempfile if needed.
+    :param timeout:          The total time to run the test for.
+    :param tempfile_kwargs:  Kwargs passed to tempfile.mkstemp() if tempfile
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Backticks (here and other places in the docs (if any), don't want to clutter the comments with the same thing)

File: coalib/testing/BearTestHelper.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+from unittest.case import skip, skipIf
+
+
+def generate_skip_decorator(bear):
+    """
+    Creates a skip decorator for a `unittest` module test from a bear.
+
+    `check_prerequisites` is used to determine a test skip.
```
Comment (hypothesist) [inline coalib/testing/BearTestHelper.py]: We use RST, so double backticks instead of single

File: coalib/testing/LocalBearTestHelper.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+                              list,
+                              msg="The given results are not a list.")
+
+        with prepare_file(lines, filename,
+                          force_linebreaks=force_linebreaks,
+                          create_tempfile=create_tempfile,
+                          tempfile_kwargs=tempfile_kwargs) as (file, fname), \
+                execute_bear(local_bear, fname, file) as bear_output:
+            msg = ("The local bear '{}' doesn't yield the right results. Or the"
+                   " order may be wrong.".format(local_bear.__class__.__name__))
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Could we personalize this message depending on the if condition below?

File: coalib/testing/LocalBearTestHelper.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+                      settings={},
+                      force_linebreaks=True,
+                      create_tempfile=True,
+                      timeout=None,
+                      tempfile_kwargs={}):
+    """
+    Generates a test for a local bear by checking the given valid and invalid
+    file contents. Simply use it on your module level like:
+
+    YourTestName = verify_local_bear(YourBear, (['valid line'],),
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Are these doctests? If yes, use `>>>`

File: coalib/testing/LocalBearTestHelper.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+            if (hasattr(bear, 'stdout_output') and
+                    isinstance(bear.stdout_output, collections.Iterable)):
+                bear.debug("Stdout:\n", "".join(bear.stdout_output))
+            if (hasattr(bear, 'stderr_output') and
+                    isinstance(bear.stderr_output, collections.Iterable)):
+                bear.debug("Stderr:\n", "".join(bear.stderr_output))
+        msg = []
+        while not bear.message_queue.empty():
+            msg.append(bear.message_queue.get().message)
+        raise AssertionError(str(err) + " \n" + "\n".join(msg))
```
Comment (hypothesist) [inline coalib/testing/LocalBearTestHelper.py]: Why the whitespace before the `\n`?

File: tests/testing/LocalBearTestHelperTest.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        with self.assertRaises(AssertionError), execute_bear(
+                self.uut_normal,  "Luke", files[0]) as result:
+            pass
+
+        with self.assertRaises(AssertionError), execute_bear(
+                self.uut_lint,  "Ben", files[0]) as result:
+            pass
+
+        with self.assertRaises(AssertionError), execute_bear(
+                self.uut_lint_with_commands,  "Anikin", files[0]) as result:
```
Comment (hypothesist) [inline tests/testing/LocalBearTestHelperTest.py]: It's `Anakin` :P

File: tests/testing/LocalBearTestHelperTest.py (hunk 3 above)
Comment (abhsag24) [inline tests/testing/LocalBearTestHelperTest.py]: there is no `setUp` function

File: coalib/testing/LocalBearTestHelper.py (hunk 9 above)
Comment (abhsag24) [inline coalib/testing/LocalBearTestHelper.py]: i guess this is just an example

File: tests/testing/LocalBearTestHelperTest.py (hunk 3 above)
Comment (hypothesist) [inline tests/testing/LocalBearTestHelperTest.py]: https://github.com/coala-analyzer/coala/pull/2293/files#diff-ca216335ff62d78f276a93743cd88ff7R24

Initialize `files` as `self.files` and use that in lines 34, 38, 42?

Comment (abhsag24) [inline tests/testing/LocalBearTestHelperTest.py]: but the above tests, wont work inside the setUp function! , `verify_local_bear` won't work

**Comment by AbdealiJK** [comment]
Bump, would be nice to get this merged.

**Comment by abhsag24** [comment]
the file has changed since i'll make a new PR with a new file

**Comment by sils1297** [comment]
@abhsag24 any news?

**Comment by Adrianzatreanu** [comment]
Bluntly assigning people that show interest/reviews to this PR so we can get it merged.

File: coalib/testing/BearTestHelper.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
```
Comment (Adrianzatreanu) [inline coalib/testing/BearTestHelper.py]: bump @adtac 's comment as it has not been addressed yet :D

**Comment by abhsag24** [comment]
This commit is about moving the file, we'll have to change the file in
later commits or open issues once the move is complete

**Comment by sils** [comment]
@abhsag24 can you pick this up again? It should be quick and easy.

**Comment by adtac** [comment]
@abhsag24 feel free to delete my previous review comments if they've been taken care of, if you feel they're just taking up space.

File: coalib/testing/BearTestHelper.py (hunk 12 above)
Comment (abhsag24) [inline coalib/testing/BearTestHelper.py]: removed those comments since they were about the file and not the movement of those files , rebasing and pushing

Comment (abhsag24) [inline coalib/testing/BearTestHelper.py]: i'll open newcomer issues once merged

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2293/commits/328bce6f20b9683487686b84821243d249e08bf9

**Comment by sils** [comment]
@rultor merge

**Pull request merged**